### ISSUE TITLE
Soliciting Feedback

### ISSUE THREAD

**Issue by julienchastang** [Unidata/thredds-docker#34]
If possible, I would like feedback from @kwilcox, @lesserwhirls, @cwardgar, @rsignell-usgs on this project which I am hoping will be the canonical TDS and TDM Docker images. I believe the merging of the Axiom and Unidata efforts concerning a TDS and TDM Docker containers should be nearing completion.

**Comment by rsignell-usgs** [comment]
Bringing @daf in here also.

**Comment by julienchastang** [comment]
And also @JohnLCaron. BTW, I forgot to mention I have a demonstration TDS/TDM container running on Azure: http://unicloud.westus.cloudapp.azure.com/thredds/catalog.html. Note, I am only brining in a subset of the data represented by the catalog tree so many nodes will be empty, but you can find a good deal of GFS data, for example:

  - http://unicloud.westus.cloudapp.azure.com/thredds/catalog/grib/NCEP/GFS/Global_onedeg/catalog.html
  - http://unicloud.westus.cloudapp.azure.com/thredds/catalog/grib/NCEP/GFS/Global_0p25deg/catalog.html 

You can bring these data into the IDV or Python for further experimentation...

**Comment by gajowi** [comment]
I have some feedback. We are moving to use this thredds-docker for our services and have one instance up for internal development needs but if all goes well, will move to use it in production.

feedback 1. the project/repository on docker hub should be more visible.  I had trouble finding that it supersedes the axiom one (that is now my understanding, can you confirm?). Naming the docker hub and github projects differently may be problematic.  Perhaps it would help to have https://hub.docker.com/r/axiom/docker-thredds/ (which has clearly been popular - and was set up nice and cleanly) point to unidata/tds. The axiom setup is also more complete on docker hub with pointers to the Dockerfile https://hub.docker.com/r/axiom/docker-thredds/~/dockerfile/ build details https://hub.docker.com/r/axiom/docker-thredds/builds/ and source repository - that makes it attractive to people looking for an existing docker thredds/tds solution.  It also helps that it come up in a search fro 'thredds' which is a more specific term than 'tds'.

feedback 2. The versioning scheme seems fine but is a little unclear to me.  I'm establishing a devops methodology (learning as I go) to automate build and deploy of containers, and am currently thinking I should have some indirection rather than pulling the images from docker hub, for a couple of reasons (security/performance/provenance). I'm intending on cloning https://github.com/Unidata/thredds-docker.git and checking out a specific tag or branch before building the image myself to deploy.  It would help to understand the versioning scheme. Perhaps the README could be extended or a separate versioning document written.  I guess there are multiple contexts where versioning is important and this complicates things, combined with the legacy of having an existing setup to maintain. The current docker hub images seem to be identical for docker hub tags 'latest and '4.6.6' (but I can't easily tell). Is 'latest' built regularly from the head of master? 4.6.6 too or is it from the 4.6.6 git tag?

feedback 3. chown -R tomcat:tomcat "${CATALINA_HOME}"  is biting me.  It might help to only chown things that need to be written to.  I guess that is just the log directories and cache.  I'm mapping a volume read-only from a read-only NFS mount on the host system (doubly read-only - can't hurt!)

**Comment by julienchastang** [comment]
@gajowi First, thanks for the valuable feedback.

> I have some feedback. We are moving to use this thredds-docker for our services and have one instance up for internal development needs but if all goes well, will move to use it in production.

Thanks. This is good news.

> feedback 1. the project/repository on docker hub should be more visible. I had trouble finding that it supersedes the axiom one (that is now my understanding, can you confirm?). 

I am hoping that is the ultimate aim, but we are still all working to attain that objective.

> Naming the docker hub and github projects differently may be problematic. Perhaps it would help to have https://hub.docker.com/r/axiom/docker-thredds/ (which has clearly been popular - and was set up nice and cleanly) point to unidata/tds. 

@lesserwhirls had some thoughts about this subject as well as others he will be providing in a separate reply. Also, I would like the name to be consistent with the rest of our docker portfolio (e.g., https://github.com/Unidata/ramadda-docker, https://github.com/Unidata/ldm-docker, etc.)

> The axiom setup is also more complete on docker hub with pointers to the Dockerfile https://hub.docker.com/r/axiom/docker-thredds/~/dockerfile/ build details https://hub.docker.com/r/axiom/docker-thredds/builds/ and source repository - that makes it attractive to people looking for an existing docker thredds/tds solution. It also helps that it come up in a search fro 'thredds' which is a more specific term than 'tds'.

Clearly the unidata/tds dockerhub page needs to be more fleshed out, as you describe. Moreover, the github repo needs to point back to dockerhub which currently it does not. I will open issues concerning these problems in a moment. Though I would strongly prefer to not manually duplicate verbage between dockerhub and github. I would either like to point back to github for the documentation or perhaps dockerhub offers an elegant way to reference Docker related files  on github. (A long time ago, I looked into this, but had no luck, but hopefully my information is out of date.)

> feedback 2. The versioning scheme seems fine but is a little unclear to me.

See next comment:

>  I'm establishing a devops methodology (learning as I go) to automate build and deploy of containers, and am currently thinking I should have some indirection rather than pulling the images from docker hub, for a couple of reasons (security/performance/provenance). I'm intending on cloning https://github.com/Unidata/thredds-docker.git and checking out a specific tag or branch before building the image myself to deploy. It would help to understand the versioning scheme. Perhaps the README could be extended or a separate versioning document written. I guess there are multiple contexts where versioning is important and this complicates things, combined with the legacy of having an existing setup to maintain. The current docker hub images seem to be identical for docker hub tags 'latest and '4.6.6' (but I can't easily tell). Is 'latest' built regularly from the head of master? 4.6.6 too or is it from the 4.6.6 git tag?

OK, I know what you mean by the mismatch between git tags and container versions. So far I have not been diligent about keeping these in sync. Pretty much the container versioning scheme on dockerhub corresponds to releases on the Unidata Maven artifact server, but does not match the git tags, unfortunately. In sum, there are three items to keep in sync: TDS versions, thredds-docker repo versions, and dockerhub container  versions. I am not sure that any errors made in the past can be rectified at this point, but going forward we can certainly try to be more diligent here. Thanks for bringing attention to this important versioning matter.

BTW, the dockerhub `latest` and `4.6.6` container version are identical. `latest` does not correspond to the HEAD of master on `thredds-docker`. 

> feedback 3. chown -R tomcat:tomcat "${CATALINA_HOME}" is biting me. It might help to only chown things that need to be written to. I guess that is just the log directories and cache. I'm mapping a volume read-only from a read-only NFS mount on the host system (doubly read-only - can't hurt!)

We need to better follow security best practices in this area, [1] [2] and that `chown` does indeed look fishy. I'll open up an issue about this too.

[1] https://www.owasp.org/index.php/Securing_tomcat
[2] https://tomcat.apache.org/tomcat-7.0-doc/security-howto.html

**Comment by lesserwhirls** [comment]
Just a few points from the discussion during the THREDDS meeting today:

1. What does the axiom tomcat container buy us over the official tomcat container? I do see it has some extra features, but is it worth the maintenance of a separate tomcat container when it comes to the TDS?

2. When it comes to locking down tomcat, see our recomendataions [here](https://www.unidata.ucar.edu/software/thredds/current/tds/tutorial/Security.html). Speficially, when it comes to permissions inside of `${CATALINA_HOME}`, see [here](https://www.unidata.ucar.edu/software/thredds/current/tds/tutorial/Security.html#install).

3. Soon we will be hosting our artifacts on the Unidata Nexus server, so any reference to the ftp server at Unidata should be replaced with refernces to https://artifacts.unidata.ucar.edu/content/repositories/unidata-releases/edu/ucar/ or https://artifacts.unidata.ucar.edu/content/repositories/unidata-snapshots/edu/ucar/
and use either the artifacts under the tds or tdmFat modules.

4. @DennisHeimbigner mentioned using a keystore for the ssl cert, but I cannot recall the exact reason at the moment (I think there is the possibility of using other certs with TDS?).

**Comment by lesserwhirls** [comment]
HI @julienchastang - you are correct, TDS does capture things more accuratly, as THREDDS is the project, and netCDF-Java and the TDS (and now Rosetta) are technologies produce from the project. Then again, when most people talk about THREDDS, more often than not they are talking about the data server, so there is a common perception that THREDDS <--> TDS (dang branding). Given that we would not likely put toolsUI or netCDF-Java in a container for others to use (maybe it does make sense to maintain one for netCDF-Java, even though it would be extremely light weight?), maybe using the name thredds is the way to go...then again, Rosetta will have its own container, and it is now under the THREDDS umbrella.

I'd say:

netCDF-Java, toolsUI -> likely won't have a container unless a strong case can be made
TDS -> call the TDS container stuff THREDDS, but when you can, specifically call out that this is the data server.
Rosetta -> the container will just be Rosetta.

**Comment by kwilcox** [comment]
> 1.  What does the axiom tomcat container buy us over the official tomcat container? I do see it has some extra features, but is it worth the maintenance of a separate tomcat container when it comes to the TDS?

The default `tomcat` container was insecure and in dire need of a maintainer when I wrote `axiom/docker-tomcat`. I haven't revisted it since but if there are no benefits we should switch to the default `tomcat8`.

> If possible, I would like feedback from @kwilcox, @lesserwhirls, @cwardgar, @rsignell-usgs on this project which I am hoping will be the canonical TDS and TDM Docker images

I've never seen docker-compose used as a replacement for `docker run` like this. I think it could be better documented to show "composing" together a `tds` and a `tdm` container so that they may interact over the internal docker networks (or any network the person managing docker uses). Right now it's just showing how to use docker-compose as a short-cut to lengthy `docker run` commands. If we are doing to explain in the README how to use `docker ps`, we should just include the actual `docker run` commands so someone knows what they are doing.

The "automated build" capabilities of Docker Hub should allow you to avoid duplication when it comes to documentation. It will pull in the README and link back to this repository. I recommend doing it this way to avoid mis-matched documentation headaches.

I have `axiom/docker-thredds` setup to build each `tag` as an image of the same name, and each `branch` as a specific minor version (4.6, 5.0, etc.). I was choosing a stable branch (4.6) to build as `latest`.

![selection_008](https://cloud.githubusercontent.com/assets/13939/18296663/2c973bda-7479-11e6-88b0-022340aae426.png)

In summary, it makes sense to me to have three repositories... `tds-docker`, `tdm-docker` and `thredds-docker`.

* `tds-docker` - What was in `axiom/docker-thredds` with enhancements that happened when it moved here
* `tdm-docker` - A new repo to manage builds/issues regarding the `tdm`.
* `thredds-docker` - Documentation and samples on how to use the `tds-docker` and `tdm-docker` images along with `docker-compose` to stand up fully features THREDDS installations.

**Comment by julienchastang** [comment]
@kwilcox Thanks for the feedback

> The default tomcat container was insecure and in dire need of a maintainer when I wrote axiom/docker-tomcat. I haven't revisted it since but if there are no benefits we should switch to the default tomcat8.

I'll be addressing this topic in #41 with additional forthcoming discussion. 

> I've never seen docker-compose used as a replacement for docker run like this. I think it could be better documented to show "composing" together a tds and a tdm container so that they may interact over the internal docker networks (or any network the person managing docker uses). Right now it's just showing how to use docker-compose as a short-cut to lengthy docker run commands. If we are going to explain in the README how to use docker ps, we should just include the actual docker run commands so someone knows what they are doing.

I'll address in this #43

> The "automated build" capabilities of Docker Hub should allow you to avoid duplication when it comes to documentation. It will pull in the README and link back to this repository. I recommend doing it this way to avoid mis-matched documentation headaches.

Thanks for expanding upon "automated builds". I did not know about this feature. I'll address this in #44.

> In summary, it makes sense to me to have three repositories... tds-docker, tdm-docker and thredds-docker.

I cannot tell here if you are talking github or dockerhub repos (and if this is related to "automated builds"). If this concerns github, I would not be in favor of setting up so many repositories. Especially because the TDM and TDS are closely linked. I'll have to further study best practices in this area.
